boards/shared/Test/Src/Test_SharedStateMachine.cpp

```diff
@@ -0,0 +1,115 @@
+#include <gtest/gtest.h>
+#include <fff.h>
+
+#include "Test_Utils.h"
+#include "App_TestWorld.h"
+
+extern "C"
+{
+#include "App_SharedStateMachine.h"
+}
+
+
+FAKE_VOID_FUNC(state_A_entry, struct StateMachine *);
+FAKE_VOID_FUNC(state_A_tick_1kHz, struct StateMachine *);
+FAKE_VOID_FUNC(state_A_exit, struct StateMachine *);
+FAKE_VOID_FUNC(state_B_entry, struct StateMachine *);
+FAKE_VOID_FUNC(state_B_tick_1Hz, struct StateMachine *);
+FAKE_VOID_FUNC(state_B_tick_1kHz, struct StateMachine *);
+FAKE_VOID_FUNC(state_B_exit, struct StateMachine *);
+FAKE_VOID_FUNC(state_C_entry, struct StateMachine *);
+FAKE_VOID_FUNC(state_C_exit, struct StateMachine *);
+
+static struct State state_A;
+static struct State state_B;
+static struct State state_C;
+
+class SharedStateMachineTest : public testing::Test
+{
+  protected:
+    void SetUp() override
+    {
+        world = App_TestWorld_Create();
+
+        state_A.run_on_entry     = state_A_entry;
+        state_A.run_on_tick_1Hz  = state_A_tick_1Hz;
+        state_A.run_on_tick_1kHz = state_A_tick_1kHz;
+        state_A.run_on_exit      = state_A_exit;
+
+        state_B.run_on_entry     = state_B_entry;
+        state_B.run_on_tick_1Hz  = state_B_tick_1Hz;
+        state_B.run_on_tick_1kHz = state_B_tick_1kHz;
+        state_B.run_on_exit      = state_B_exit;
+
+        state_C.run_on_entry     = state_C_entry;
+        state_C.run_on_tick_1Hz  = NULL;
+        state_C.run_on_tick_1kHz = NULL;
+        state_C.run_on_exit      = state_C_exit;
+
+        state_machine = App_SharedStateMachine_Create(world, &state_A);
+
+        RESET_FAKE(state_A_entry);
+        RESET_FAKE(state_A_tick_1kHz);
+        RESET_FAKE(state_A_exit);
+        RESET_FAKE(state_B_entry);
+        RESET_FAKE(state_B_tick_1Hz);
+        RESET_FAKE(state_B_tick_1kHz);
+        RESET_FAKE(state_B_exit);
+    }
+
+    void TearDown() override
+    {
+        TearDownObject(world, App_TestWorld_Destroy);
+        TearDownObject(state_machine, App_SharedStateMachine_Destroy);
+    }
+
+    void SetInitialState(const struct State *const initial_state)
+    {
+        App_SharedStateMachine_Destroy(state_machine);
+        state_machine = App_SharedStateMachine_Create(world, initial_state);
+        ASSERT_TRUE(state_machine != NULL);
+        ASSERT_EQ(
+            initial_state,
+            App_SharedStateMachine_GetCurrentState(state_machine));
+    }
+
+    // We provide our own implementation of the 1hz tick for state_A
+    // here so that we can simulate a state transition in a tick
+    static void state_A_tick_1Hz(struct StateMachine *state_machine)
+    {
+        App_SharedStateMachine_SetNextState(state_machine, &state_B);
+    }
+
+    struct World *       world;
+    struct StateMachine *state_machine;
+};
+
+TEST_F(
```

Comment:
Maybe add another basic test that ticks a NULL as  well as a non-NULL function, and assert on the `call_count`?

Reply:
Sorry, could you be a bit more specific in terms of what functionality you're trying to test here that's not covered in the existing tests?

Reply:
Oh just checking that we don't tick if it's NULL

Reply:
What exactly would you be asserting? If the tick function is NULL we can't very well check that it's never called, can we?

Reply:
Oh yeah of course, whoops :facepalm: